src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java

```diff
@@ -0,0 +1,355 @@
+                if (line.contains(" " + command)) {
+                    pid = Integer.parseInt(line.trim().replaceAll("(\\d+).+", "$1"));
+                    pids.add(pid);
+                }
+            }
+        });
+        return pids;
+    }
+
+    // Kthreadd process (pid = 1) should have 52 fields in its Stat file, and it should have (systemd) as comm value and 1 as pid value in the file. None of the fields should be null
+    @Test
+    public void readSystemdFileTest() {
+        Assertions.assertDoesNotThrow(() -> {
+            systemd = new Process(1);
+            Stat stat = systemd.stat();
+            Assertions.assertEquals(1, stat.pid());
+            Assertions.assertEquals("(systemd)", stat.comm());
+            Iterator iterator = stat.read().iterator();
+            int i = 0;
+            while (iterator.hasNext()) {
+                Assertions.assertNotNull(iterator.next());
+                i++;
+            }
+            Assertions.assertEquals(52, i);
+
+        });
+        systemd = new Process(1);
+    }
+
+    // Kthreadd process (pid = 2) should have 52 fields in its Stat file, and it should have (kthreadd) as comm value and 2 as pid value in the file. None of the fields should be null
+    @Test
+    public void readKthreaddFileTest() {
+        Assertions.assertDoesNotThrow(() -> {
+            kthreadd = new Process(2);
+            Stat stat = kthreadd.stat();
+            Assertions.assertEquals(2, stat.pid());
+            Assertions.assertEquals("(kthreadd)", stat.comm());
+            Iterator iterator = stat.read().iterator();
+            int i = 0;
+            while (iterator.hasNext()) {
+                Assertions.assertNotNull(iterator.next());
+                i++;
+            }
+            Assertions.assertEquals(52, i);
+
+        });
+    }
+
+    // Process object should be able to be instantiated with an integer and a String
+    @Test
+    public void constructorTest() {
+        Process process = new Process(1);
+        Process processString = new Process("1");
+        Assertions.assertTrue(process.isAlive());
+        Assertions.assertTrue(processString.isAlive());
+    }
+
+    // If a process with the designated ID is not found, the returned Process object should be a stub, and isAlive should return false.
+    @Test
+    public void noSuchProcessTest() {
+        Process process = new Process(-1);
+        Assertions.assertFalse(process.isAlive());
+    }
+
+    // After starting a sleeping process in a new thread, isAlive should return true. After killing the process, it should return false.
+    @Test
+    public void isAliveTest() {
+        Assertions.assertDoesNotThrow(() -> {
+            ProcessBuilder pb = new ProcessBuilder();
+            pb.command("sleep", "10");
+
+            TestThread processThread = new TestThread(pb);
+            processThread.run();
+
+            ArrayList<Integer> pids = getPidByCommand("sleep");
```

Comment:
There can be other sleeps running on the system at the same time so this is can be fairly fragile test

Reply:
Preferably this test would create a process, make sure that isAlive() is true, then kill the process and verifiy that isAlive is now false. If there are other sleeps running, this test will fail, and worse yet, it will kill the sleep process it finds (even if it's someone else's process).

This could be solved if we knew the process ID of the created process, but java.lang.Process.pid() is a Java 9 feature. There seems to be some ways to get it via reflection, but at least in my tests I've not been able to access the process id field within Java's ProcessImpl easily.

Is there some way we can create a process where we are able to identify it properly without knowing its pid?
Or should we just test isAlive() by calling it on the JVM process, verifying that it returns true, and leaving it at that?
At least we shouldn't be killing sleep processes that have been started somewhere else.

Reply:
I think I managed to solve this:
We don't need to kill "sleep", since it terminates itself after the given time, we can just wait for the duration of the sleep process before checking isAlive() again.
Added a check that makes sure there's only one sleep process active. If there's more than one, the test will simply fail. Otherwise we might get false positives.